### PR TITLE
Implement non-bool results from (in)equality tests

### PR DESCRIPTION
**Pull request by nickovs**
This PR implements a more complete replication of CPython's behaviour for equality and inequality testing. This addresses the issues discussed in #5382 and a few other warts. Improvements over the old code include:
 * Support for returning non-boolean results from comparisons (as used by `nunmpy` and others).
 * Support for non-reflexive equality tests.
 * Preferential use of `__ne__` methods and `MP_BINARY_OP_NOT_EQUAL` binary operators for inequality tests, when available.
 * Fallback to `op2 == op1` or `op2 != op1` when `op1` does not implement the (in)equality operators.

The PR also includes a more comprehensive set of tests for the behaviour of (in)equality operators implemented in special methods.

### CODE REVIEW

**Comment by nickovs** [comment]
This new code does appear to make things bigger (and that is what Travis is complaining about). Notably, on the "bare-arm and minimal ports build" the code size currently goes from 68532 bytes to 68672 bytes (an extra 140 bytes, or 0.2%). I'm sure that we can make some space/time trade-offs here if we are willing to forego some of the optimisations that fast-path common tests, but that could impact performance. It might be possible to make these trade-offs configurable in `mpconfigport.h`.

@dpgeorge What price are you willing to pay to be compatible with CPython in this area?

**Comment by dpgeorge** [comment]
@nickovs if you could rebase this on to the latest master (which includes your slice changes) that would help to see how the code size changes.

**Comment by nickovs** [comment]
Well there's a first: I managed to get `git rebase` to do what I wanted it to do first time!

@dpgeorge I have squashed all the commits into one, rebased against the current master. Hopefully this will make your comparison a little easier.

File: py/obj.c [hunk 1]
```diff
@@ -192,7 +189,7 @@ bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
     if (mp_obj_is_str(o1)) {
         if (mp_obj_is_str(o2)) {
             // both strings, use special function
-            return mp_obj_str_equal(o1, o2);
+            return mp_obj_str_equal(o1, o2) ? local_true : local_false;
         } else {
             // a string is never equal to anything else
```
Comment (nickovs) [inline py/obj.c]: Note that _'a string is never equal to anything else'_ is incompatible with CPython, since a custom class can decide that an instance is equal to a string. Changing this to something like:
```C
    if (mp_obj_is_str(o1) && mp_obj_is_str(o2)) {
        return mp_obj_str_equal(o1, o2) ? local_true : local_false;
    }
```
would make the code smaller and more compatible, at the expense of some speed for unusual comparisons, although it would also require some fix-ups in `objstr.c` since that code has some assumptions about not being passed strings in comparisons.

**Comment by dpgeorge** [comment]
Thanks for rebasing.  Code size change for this PR is:
```
   bare-arm:  +120 +0.181% 
minimal x86:  +256 +0.173% 
   unix x64:  +256 +0.051% 
unix nanbox:  +320 +0.072% 
      stm32:  +100 +0.026% PYBV10
     cc3200:  +128 +0.069% 
    esp8266:  +136 +0.020% GENERIC
      esp32:   +16 +0.001% GENERIC[incl +16(data)]
        nrf:  +120 +0.082% pca10040
       samd:  +112 +0.110% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

I'll need to review it in detail, but in general I think it's a change worth paying for (in terms of code size).

File: py/obj.c [hunk 2]
```diff
@@ -159,28 +159,25 @@ bool mp_obj_is_callable(mp_obj_t o_in) {
...
+    mp_obj_t local_true = not_equal ? mp_const_false : mp_const_true;
+    mp_obj_t local_false = not_equal ? mp_const_true : mp_const_false;
+
+    // Shortcut for very common cases
+    if (o1 == o2 &&
+        (mp_obj_is_small_int(o1) || o1 == mp_const_none) ) {
+        return local_true;
     }
+
     if (o1 == mp_const_none || o2 == mp_const_none) {
```
Comment (dpgeorge) [inline py/obj.c]: I think this if statement needs to be removed (or changed somehow...) because it's up to a user type if it's equal to None or not.

File: py/obj.c [hunk 3]
```diff
@@ -159,28 +159,25 @@ bool mp_obj_is_callable(mp_obj_t o_in) {
...
-    if (o1 == o2
-        #if MICROPY_PY_BUILTINS_FLOAT
-        && !mp_obj_is_float(o1)
-        #endif
-        #if MICROPY_PY_BUILTINS_COMPLEX
-        && !mp_obj_is_type(o1, &mp_type_complex)
-        #endif
-        ) {
-        return true;
+mp_obj_t mp_obj_equal_bop(mp_obj_t o1, mp_obj_t o2, bool not_equal) {
```
Comment (dpgeorge) [inline py/obj.c]: I tried changing this function signature to:
```c
mp_obj_t mp_binary_op_equal(mp_binary_op_t op, mp_obj_t o1, mp_ob_t o2) {
    mp_obj_t local_true = (op == MP_BINARY_OP_NOT_EQUAL) ? mp_const_false : mp_const_true;
    ...
}
```
and it made the code size smaller by about 16 bytes on almost all ports.  And IMO it's a bit neater this way because it mirrors `mp_binary_op()`, just a special version for equal/not equal.

File: py/obj.c (hunk 1 above)
Comment (dpgeorge) [reply py/obj.c]: Yeah.  Might be worth doing this in a separate PR/commit, similar to how None is treated (and maybe also int?).

File: py/obj.c [hunk 4]
```diff
@@ -205,21 +202,41 @@ bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
...
 
-    // equality not implemented, and objects are not the same object, so
-    // they are defined as not equal
-    return false;
+    // equality not implemented, so fall back to pointer conparison
+    return (o1 == o2) ? local_true : local_false;
+}
+
+bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
+    return mp_obj_is_true(mp_obj_equal_bop(o1, o2, false));
```
Comment (dpgeorge) [inline py/obj.c]: This is the function that matters for performance, because it's used in things like dict lookup and sequence comparison.  Prior to the change here it was "fast" because it was just returning true/false (C values) directly.  But now it must go through `mp_obj_is_true()` to convert the object to a C-level boolean.  Of course this is more powerful (and correct wrt to CPython).  To get back prior performance (if there was any) probably need to inline part of `mp_obj_is_true` here (eg the test for `mp_const_true`/`mp_const_false`).  I'm not saying this needs to be done, just considering the impact.

File: py/runtime.c [hunk 5]
```diff
@@ -323,19 +323,8 @@ mp_obj_t mp_binary_op(mp_binary_op_t op, mp_obj_t lhs, mp_obj_t rhs) {
...
-            }
-        } else {
-            if (op == MP_BINARY_OP_EQUAL) {
-                return mp_const_false;
-            } else {
-                return mp_const_true;
-            }
-        }
+        // mp_obj_equal_bop supports a bunch of shortcuts
+        return mp_obj_equal_bop(lhs, rhs, op == MP_BINARY_OP_NOT_EQUAL);
```
Comment (dpgeorge) [inline py/runtime.c]: With the change to signature suggested above, this becomes simply `mp_binary_op_equal(op, lhs, rhs)`

**Comment by dpgeorge** [comment]
Measuring performance change on PYBv1.0 I get (before to after):
```
diff of scores (higher is better)
N=100 M=100                    pyb0 ->       pyb1         diff      diff% (error%)
bm_chaos.py                  297.27 ->     296.11 :      -1.16 =  -0.390% (+/-0.01%)
bm_fannkuch.py                73.91 ->      73.44 :      -0.47 =  -0.636% (+/-0.00%)
bm_fft.py                   2447.58 ->    2448.13 :      +0.55 =  +0.022% (+/-0.00%)
bm_float.py                 4791.08 ->    4827.77 :     +36.69 =  +0.766% (+/-0.01%)
bm_hexiom.py                  35.20 ->      34.82 :      -0.38 =  -1.080% (+/-0.00%)
bm_nqueens.py               4205.63 ->    4109.67 :     -95.96 =  -2.282% (+/-0.00%)
bm_pidigits.py               636.72 ->     635.29 :      -1.43 =  -0.225% (+/-0.25%)
core_with.py               45226.10 ->   44972.62 :    -253.48 =  -0.560% (+/-0.00%)
core_yield_from.py         58252.15 ->   58081.71 :    -170.44 =  -0.293% (+/-0.04%)
misc_aes.py                  355.52 ->     357.61 :      +2.09 =  +0.588% (+/-0.01%)
misc_mandel.py              2990.19 ->    2963.71 :     -26.48 =  -0.886% (+/-0.01%)
misc_pystone.py             1926.47 ->    1936.32 :      +9.85 =  +0.511% (+/-0.00%)
misc_raytrace.py             303.38 ->     302.70 :      -0.68 =  -0.224% (+/-0.01%)
viper_call0.py               576.78 ->     576.79 :      +0.01 =  +0.002% (+/-0.00%)
viper_call1a.py              550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py              434.78 ->     434.87 :      +0.09 =  +0.021% (+/-0.02%)
viper_call1c.py              439.35 ->     439.44 :      +0.09 =  +0.020% (+/-0.00%)
viper_call2a.py              536.23 ->     536.26 :      +0.03 =  +0.006% (+/-0.01%)
viper_call2b.py              377.17 ->     377.25 :      +0.08 =  +0.021% (+/-0.00%)
```
There's a 0.5% increase in performance on misc_aes and misc_pystone, but 1% and 2.3% decrease for bm_hexiom and bm_nqueens.  The rest are very small changes.  It's not much of a increase/decrease and it's not really clear how the PR here actually affected the change in performance.  Would need to dig in to each test and see in detail what happened.  But that's probably not a great use of time, better to have correct behaviour first (ie this PR), then optimise later if needed.

**Comment by dpgeorge** [comment]
Some more tests (specifically for handling None and strs, which currently fail with this PR):
```python
    def __eq__(self, other):
        print('A eq', other)
        return 123

class B:
    def __ne__(self, other):
        print('A ne', other)
        return -456

print(A() != B())
print(B() != A())

for val in (None, 0, 1, 'a'):
    print('==== testing', val)
    print(A() == val)
    print(val == A())
    print(A() != val)
    print(val != A())
    print(B() == val)
    print(val == B())
    print(B() != val)
    print(val != B())
```

File: py/obj.c (hunk 3 above)
Comment (nickovs) [reply py/obj.c]: Makes sense. I'll make that change.

File: py/obj.c (hunk 2 above)
Comment (nickovs) [reply py/obj.c]: Quite right. It *should* just work if I take this out.

File: py/obj.c (hunk 1 above)
Comment (nickovs) [reply py/obj.c]: Right. I will leave this as is until this PR is merged and then sort out the fixes to make strings work separately.

File: py/obj.c (hunk 4 above)
Comment (nickovs) [reply py/obj.c]: Originally I implemented this in line and then reverted to this version because I was concerned about code size. I wonder if it might make sense to set up some sort of conditionally in-lining based on a setting on port configuration.

File: py/obj.c (hunk 2 above)
Comment (nickovs) [reply py/obj.c]: I check, and everything still works when simply removing this path.

**Comment by dpgeorge** [comment]
Looking further in to this digs up some more issues:
- tuple, list, dict, set all have a shortcut for `MP_BINARY_OP_EQUAL` that will return false if they are compared to something other than their own type (like it was with None, and still is with str).  Removing these shortcuts (just an if in their binary_op method) might be enough to fix it.
- Existing built-in types (like tuple, dict) all assumed reflexivity and that `==` is the inverse of `!=`, so they just had a path for `MP_BINARY_OP_EQUAL` in one direction (ie didn't try to swap args, try `MP_BINARY_OP_NOT_EQUAL`).  So with the PR here, doing a `!=` on a built-in type will first try `MP_BINARY_OP_NOT_EQUAL` then try `MP_BINARY_OP_EQUAL` (then swap args if neither works).  Trying `MP_BINARY_OP_NOT_EQUAL` -- which always fails on built-in types because it's not implemented -- probably leads to some reduced performance.
- A few places in the code explicitly say not to pass through `MP_BINARY_OP_NOT_EQUAL` (eg `mp_seq_cmp_objs`) so it's not straightforward to just pass it through to fix the above-mentioned inefficiency (if it really is an inefficiency).
- Looking in to `mp_seq_cmp_objs` to try and make it work with `MP_BINARY_OP_NOT_EQUAL` found some other problems with it, that it converts "<" to ">" (will file separate issue about this).

It's a small can of worms.  The point is that uPy's equality logic was implemented under the assumptions that "==" is strictly true or false, "!=" is the opposite of "==", and didn't really consider that user types may want to be compared to built-in types.

For most Python applications these are fine assumptions, and can lead to a "micro" implementation of Python (ie MicroPython) which doesn't have full support for the object model and rich comparison.  So I guess the main question is whether we want to break these assumptions and make it all work as per (C)Python's semantics.

I'd say it's worth trying to make it work.  It's mostly done here and just needs a bit of further work.  To start with I'd suggest just removing the lines like this one: https://github.com/micropython/micropython/blob/master/py/objtuple.c#L114-L116

From an efficiency point of view, I think the main things to optimise are when the comparison is between two objects with the same type (eg list and list).  And I think that would already be well optimised, by default (so no need to do anything extra at this point).

**Comment by nickovs** [comment]
In the case of comparison of built-in types against other built-in types, the shortcuts, reflexivity, inevitability etc. can be known in advance. For comparisons of user-defined types against built-in types or other user-defined types all bets are off.

One way to deal with this would assign some bits in the class's flags (half)word to indicate which of the short-cuts can be applied and have it fall through to the generic code if no flags are set (and not set them for user-defined types). This would clean up all of the checking if the special cases are applicable (except for small ints), which might make the code smaller, while keeping all of the `==`/`!=` symmetry and reflexivity code in one place.

I can see three flags that we might want to have:
 * Reflexive (which would probably be set for all built-in types). This means that `o1 == o2` implies `true`.
 * No `_NOT_EQUAL` op, so use the `_EQUAL` op instead (and implicitly, _EQUAL only yields boolean values).
 * Can't be equal to any other built-in type, so don't try (although it can still be equal to a custom type).

Testing for these bits in `o1->flags` and occasionally `o2->flags` would be quicker and smaller than checking the object class against at stack of pointers to built-in classes and hopefully it doesn't need much code rework outside the functions we are already fixing.

**Comment by dpgeorge** [comment]
> Testing for these bits in `o1->flags` and occasionally `o2->flags` would be quicker and smaller than checking the object class against at stack of pointers to built-in classes and hopefully it doesn't need much code rework outside the functions we are already fixing.

Or it might be possible to just check if the object type is a user type, and if not then assume it is built-in with all the standard assumptions about equality (only require one check).

**Comment by nickovs** [comment]
> Or it might be possible to just check if the object type is a user type, and if not then assume it is built-in with all the standard assumptions about equality (only require one check).

That depends on what you mean by "user type". I headed off down this whole rabbit-hole because I was implementing multi-dimensional arrays with semantics similar to `numpy`. In that case:
```
>>> a = np.array([[1,2],[3,4]])
>>> a == a
ndarray([[True, True], [True, True]])
```
This is a class written in C in a module, rather than a built-in type, so the test would need to be able to tell the difference.

**Comment by dpgeorge** [comment]
> This is a class written in C in a module, rather than a built-in type, so the test would need to be able to tell the difference.

Ok, good point.  My idea was to just use `mp_obj_is_instance_type(type)` for the check, but that won't handle the case you mention.

**Comment by dpgeorge** [comment]
> I can see three flags that we might want to have: ...

I think these can all be combined into a single flag "standard equality".  When both types have this flag set the logic can be what already exists.  If either type doesn't have it set then the general case must be handled, ie try LHS==RHS, RHS==LHS; or LHS!=RHS, !(LHS==RHS), RHS!=LHS, !(RHS==LHS).

---------------

My main concern with the PR as it currently stands is that:
- a LHS!=RHS operation on built-in types first tries LHS!=RHS (which always fails) then falls back to !(LHS==RHS); eg tuple!=tuple
- if the types are not compatible (or are compatible but don't implement ==) then the logic tries LHS/RHS swapped, which will also fail; eg tuple==list

Implementing `MP_BINARY_OP_NOT_EQUAL` would help with the first issue but would increase code size by a fair bit to handle this != in all built-in types (those which already support ==).

The second problem is harder to fix in a neat way.  Probably the only way to do it is with a flag in the type, as discussed just above.

**Comment by nickovs** [comment]
The order of the tests that I apply was used because it is what CPython does. Initially I tried reversing the arguments before attempting the inverted operator but there were a bunch of test cases that failed when I did that.

I think you are right that we can use fewer flags. All built-in types are reflexive, boolean-returning and handle inversion in the expected manner so clearly all those can be rolled into one flag. That said, the built-in numeric types can test equal to each other even when the types don't match so the question is if it is better to handle these cases with in-line code or another flag.

**Comment by mcdeoliveira** [comment]
Just wanted to support this PR. We arrived at the same issue as @nickovs for the exact same reason. Our solution is very similar but implemented with inlines for performance. It also makes no changes to the current micropython equality behavior. All tests pass without modification. Did not look at the code size issue in detail but the same comments above would apply. I updated our signatures to match those discussed above. Here is a link in case you want to check it out.

https://github.com/ComplexArts/micropython/tree/equality

I will not be submitting a PR since this one is so close to what we have.

**Comment by dpgeorge** [comment]
> Just wanted to support this PR.

@mcdeoliveira thanks for the vote of support.  There's a reworked version of this PR at #5533 which should have better performance, and same behaviour, and that PR is about to be merged.

**Comment by dpgeorge** [comment]
Superseded by #5593
